File: AlgorithmReview/Programmers/Lv2/롤케이크 자르기.swift
//
//  롤케이크 자르기.swift
//  AlgorithmReview
//
//  Created by 박소윤 on 2023/05/27.
//

import Foundation

func solution(_ topping:[Int]) -> Int {
    
    var toppingA = [Int:Int]()
    var toppingB = [Int:Int]()
    
    for i in topping {
        if toppingB[i] == nil {
            toppingB[i] = 0
        }
        toppingB[i]! += 1
    }
    
    var answer = 0
    for i in topping {
        
        if toppingA[i] == nil {
            toppingA[i] = 0
        }
        toppingA[i]! += 1
        
        if toppingB[i] == 1 {
            toppingB.removeValue(forKey: i)
        }
        else {
            toppingB[i]! -= 1
        }
        
        if toppingA.count == toppingB.count {
            answer += 1
        }
    }
    
    return answer
}

File: Python/롤케이크 자르기.py
def solution(topping):
    
    topping_a = {}
    topping_b = {}
    for i in topping:
        if i not in topping_b:
            topping_b[i] = 0
        topping_b[i] += 1
    
    answer = 0
    for i in topping:
        if i not in topping_a:
            topping_a[i] = 0
        topping_a[i] += 1
        
        if topping_b[i] == 1:
            del topping_b[i]
        else:
            topping_b[i] -= 1
            
        if len(topping_a) == len(topping_b):
            answer += 1
        
    return answer
